ratelimit_test.go

```diff
@@ -64,21 +72,28 @@ func runTest(t *testing.T, fn func(testRunner)) {
 
 	for _, tt := range impls {
 		t.Run(tt.name, func(t *testing.T) {
-			// Set a non-default time.Time since some limiters (int64 in particular) use
-			// the default value as "non-initialized" state.
-			clockMock := clock.NewMock()
-			clockMock.Set(time.Now())
 			r := runnerImpl{
 				t:           t,
-				clock:       clockMock,
+				clock:       makeTestTime(),
 				constructor: tt.constructor,
 				doneCh:      make(chan struct{}),
 			}
 			defer close(r.doneCh)
 			defer r.wg.Wait()
 
 			fn(&r)
-			r.clock.Add(r.maxDuration)
+			go func() {
+				move := func() {
+					defer func() {
+						_ = recover()
+						time.Sleep(10 * time.Millisecond)
+					}()
+					r.clock.advanceToTimer()
+				}
+				for {
+					move()
+				}
```

Comment:
Sorry, but I think this needs a much larger explanation - perhaps as comments in code?

Am I reading this right that this is an infinite loop, trying to move to infinitely move to the next timer ... by recovering from a panic?

Irrespective of goroutine leaks, using panics for signaling - do you think the `maxDuration` is not useful here? It seems we should be able to just pass the duration into `advance`, theoretically?

Reply:
> Irrespective of goroutine leaks

I think 1 goroutine leak per test case is not a big deal, but if you want, I can add some exit logic here.

>  using panics for signaling

Code from Ian just panics when you try to run the next timer, and there is nothing to run. We can potentially change that behavior to not handle panics, but I wanted to limit changes in that code.

> Sorry, but I think this needs a much larger explanation - perhaps as comments in code?

I'll write some description here first and when we clarify all questions about it, I'll move it into code comments:

```
testTime internally has a queue of all timers currently registered within the clock.
In our case, there will be timers for goroutines that try to consume permissions and timers 
for goroutines that should run test asserts. When we call testTime.advanceToTimer() it basically
shifts time to the next registered timer, runs it, and also runs runtime.Gosched() multiple times 
to give a chance for other goroutines to run and react to what happened in the current timer. 
Usually, when we run runtime.Gosched() other goroutines will observe the new state, make their moves, and potentially create new timers inside this clock to wait for some other state change. 

If testTime.advanceToTimer() panics, it means that there are no timers to run, 
and it can happen if the test is alre
```

 > Do you think the maxDuration is not useful here?

No, we can use that if we want to get out of the currently infinite loop, we can check on every iteration if testTime is already passed maxDuration.

>  It seems we should be able to just pass the duration into advance, theoretically?

Yes, but since we advance time in different goroutines in this approach, it is necessary for us to do it one step at a time and give other goroutines the opportunity to make their steps as well. There also can be situations where we start to advance time in this goroutine before test goroutines spawned all permission consumers and asserting goroutines, so we need to just wait for everything to setup, that is what we do when panic happens.

Reply:
So perhaps naively, but do we need both
- advanceToTimer
- advance(Unlocked?)

Our test runner knows exactly how much time needs to move (`maxDuration`). We can just pass it into `advance`. `advanceUnlocked` already does all the `runtime.Gosched` needed.

And soit seem to me that `advance` should theoretically be enough?
If the problem is the initial goroutine setup, should we just add the `16*runtime.Gosched` at the **beginning** of `advance` to let the scheduler kick in?
This would also help with `r.clock.advance(time.Nanosecond)` you have below.

I understand this kinda modifies the code we copy-paste, but I'd feel much more comfortable with we decreased the API by 50% here.

Reply:
@rabbbit

Right now, it is hard for me to make changes because tests do not pass until we merge https://github.com/uber-go/ratelimit/pull/95

I agree with the comments and changes that you propose right now, but they don't change the approach in general. 
I propose to merge https://github.com/uber-go/ratelimit/pull/95, and then I can apply your suggested changes, or we can merge this PR as it is right now, and I'll make these improvements in subsequent PRs.

Reply:
Merged #95. I might still want to re-review it later on, it feels like some simplification is possible.

Reply:
I'll update this branch with new version and this should potentially fix all tests, event though I'm not sure how your new test for initial calls will work 🤞